service/controller/v13/resource/instance/create.go

```diff
@@ -33,754 +39,42 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-// firstInstanceInProgress returns nil.
-func firstInstanceInProgress(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM) *compute.VirtualMachineScaleSetVM {
-	for _, v := range list {
-		if v.ProvisioningState == nil || key.IsFinalProvisioningState(*v.ProvisioningState) {
-			continue
-		}
-
-		return &v
-	}
-
-	return nil
-}
-
-// firstInstanceToReimage returns the first instance to be reimaged. The
-// decision of reimaging an instance is done by comparing the desired version
-// bundle version of the custom object and the current version bundle version of
-// the instance's tags applied. In case all instances are reimaged
-// firstInstanceToReimage return nil.
-func firstInstanceToReimage(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM, instanceNameFunc func(customObject providerv1alpha1.AzureConfig, instanceID string) (string, error), versionValue map[string]string) (*compute.VirtualMachineScaleSetVM, error) {
-	if versionValue == nil {
-		return nil, microerror.Mask(versionBlobEmptyError)
-	}
-
-	for _, v := range list {
-		desiredVersion := key.VersionBundleVersion(customObject)
-		instanceName, err := instanceNameFunc(customObject, *v.InstanceID)
-		if err != nil {
-			return nil, microerror.Mask(err)
-		}
-		instanceVersion, ok := versionValue[instanceName]
-		if !ok {
-			continue
-		}
-		if desiredVersion == instanceVersion {
-			continue
-		}
-
-		return &v, nil
-	}
-
-	return nil, nil
-}
-
-// firstInstanceToUpdate return the first instance to be updated. The decision
-// of updating an instance is done by checking if the latest scale set model is
-// applied. In case all instances are updated firstInstanceToUpdate return nil.
-func firstInstanceToUpdate(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM) *compute.VirtualMachineScaleSetVM {
-	for _, v := range list {
-		if *v.LatestModelApplied {
-			continue
-		}
-
-		return &v
-	}
-
-	return nil
-}
-
-func isNodeDrained(drainerConfigs []corev1alpha1.DrainerConfig, instanceName string) bool {
-	for _, n := range drainerConfigs {
-		if n.GetName() != instanceName {
-			continue
-		}
-		if n.Status.HasDrainedCondition() {
-			return true
-		}
-		if n.Status.HasTimeoutCondition() {
-			return true
-		}
-	}
-
-	return false
-}
-
-func (r *Resource) handleDeploymentCompletedStatus(ctx context.Context, customObject providerv1alpha1.AzureConfig) error {
-	deploymentsClient, err := r.getDeploymentsClient(ctx)
-	if err != nil {
-		return microerror.Mask(err)
-	}
-
-	d, err := deploymentsClient.Get(ctx, key.ClusterID(customObject), key.VmssDeploymentName)
-	if err != nil {
-		return microerror.Mask(err)
-	}
-
-	s := *d.Properties.ProvisioningState
-	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", s))
-
-	if key.IsSucceededProvisioningState(s) {
-		computedDeployment, err := r.newDeployment(ctx, customObject, nil)
+	if newState != currentState {
```

Comment:
Why not going with the terminology of current and desired state? What does `newState` mean?

Reply:
Because this is about transitioning from one state into another within defined state machine which logic is driven by external side-effects. State machine kind of executes logic by transitioning from one state to another. It doesn't try to work towards some desired state. In other words one could maybe say that state machine itself never stops. It's never ready if that makes sense.

You may want to also check [Finite-state machine](https://en.wikipedia.org/wiki/Finite-state_machine).

Reply:
Disagree and commit.